pkg/orchestrator/etcd_worker.go

```diff
@@ -229,58 +231,75 @@ func (worker *EtcdWorker) Run(ctx context.Context, session *concurrency.Session,
 
 		}
 
-		if len(pendingPatches) > 0 {
-			// Here we have some patches yet to be uploaded to Etcd.
-			pendingPatches, err = worker.applyPatchGroups(ctx, pendingPatches)
-			if isRetryableError(err) {
-				continue
-			}
-			if err != nil {
-				return errors.Trace(err)
-			}
-		} else {
-			if exiting {
-				// If exiting is true here, it means that the reactor returned `ErrReactorFinished` last tick,
-				// and all pending patches is applied.
-				return nil
-			}
-			if worker.revision < worker.barrierRev {
-				// We hold off notifying the Reactor because barrierRev has not been reached.
-				// This usually happens when a committed write Txn has not been received by Watch.
-				continue
+		tryCommitPendingPatches := func() (bool, error) {
+			if len(pendingPatches) > 0 {
+				// Here we have some patches yet to be uploaded to Etcd.
+				pendingPatches, commitedChanges, err = worker.applyPatchGroups(ctx, pendingPatches)
+				if isRetryableError(err) {
+					return true, nil
+				}
+				if err != nil {
+					return false, errors.Trace(err)
+				}
+				// pendingPatches size may greater than 0, but nothing to commit to etcd
+				// if nothing is changed, we should not break this tick
+				if commitedChanges > 0 {
+					return true, nil
+				}
 			}
+			return false, nil
+		}
+		if retry, err = tryCommitPendingPatches(); err != nil {
+			return err
+		}
+		if retry {
+			continue
+		}
+		if exiting {
```

Comment:
it looks, we should check `exsting` before `retry`

Reply:
done